rfcs/056-prismic-etl-pipeline/README.md

```diff
@@ -0,0 +1,91 @@
+
+## Proposal
+
+The Prismic corpus is smaller than our other corpora, and this ETL pipeline is very linear and self-contained. However, Prismic data is fairly heavily [normalised](https://en.wikipedia.org/wiki/Database_normalization) and so we need to build a solution that can (a) denormalise data from linked documents onto our "primary" documents and (b) reflect changes from these linked documents ("secondary" documents) on all of the primary documents on which they are present. 
+
+A basic example of this would be that the `article` type has a `contributors` field, which links to `role` and `person` types:
+```json
+{
+  ...
+  "contributors": {
+    "role": {
+      "id": "<foreign key>",
+      "type": "editorial-contributor-roles",
+      ...
+    },
+    "person": {
+      "id": "<foreign key>",
+      "type": "people",
+      ...
+    }
+  }
+}
+```
+
+Fortunately, Prismic provide an API, [`graphQuery`](https://prismic.io/docs/graphquery-rest-api) which can do this denormalisation for us, using a GraphQL-like syntax. This straightforwardly solves the first problem of denormalising linked data, but not the second problem of reflecting changes in it.
+
+The proposed solution to this is to store the IDs of the secondary documents alongside the other information we index in Elasticsearch for primary documents:
+```json
+{
+  "display": {
+    // Opaque JSON to be displayed by the API
+  },
+  "query": {
+    // Fields for querying, filtering etc
+    ...
+    // List of identifiers (foreign keys) of linked documents
+    "ids": [
+      "1",
+      "2",
+      "3"
+    ]
+  }
+}
+```
+Then we can build a pipeline that works as follows:
+
+1. A 'window generator' Lambda triggered on schedule, which generates a payload representing a time period/window that is sent to (2).
+2. A 'Prismic ETL' Lambda which consumes time periods (potentially half-bounded or unbounded, for complete reindexes) and then:
+   1. Queries Prismic for all documents (including denormalised data on primary documents) updated within the time window.
+   2. For all secondary documents, query the ES index for already-indexed primary documents that contain them.
+   3. Queries Prismic for all the documents (including denormalised data) from (ii) that are not part of the list returned by (i).
+   4. Transforms the resultant primary documents into JSON objects as described above.
+   5. Indexes these into an Elasticsearch cluster using the Elasticsearch JS client's bulk helpers. 
+
+For complete reindexes, it would be straightforward to trigger the Prismic ETL lambda from a local script, with a payload that covers all documents. In this case, it would also be an easy optimisation to disable steps (ii) and (iii), as all documents would be being fetched regardless.
```

Comment:
I wonder if this might be a job for Step Functions or Airflow (or maybe just a pair of similar-but-different queue-driven lambdas similar to how we split the bulk and SQS-driven paths in the Concepts pipeline).

Both paths could be independently triggered from the same window generator, then we could have one path that starts with primary documents and another that starts with secondary documents.

That way, rather than thinking of the optimisation to disable steps 2.2 and 2.3 for a full reindex, it would simply be running the normal primary document path over all documents since time immemorial.

Only the secondary document path would include steps 2.2 and 2.3

A further optimisation in 2.2 might be to query for already indexed primary documents that have not been updated since the latest secondary document changed.

Reply:
This is very useful - I think what I might suggest is that we start with a single Lambda which is modularised in a way informed by this comment, with a view that - if the pipeline needs to expand - we might want to split up the monolith into DAG-y microservices further down the line.